core/src/gossip_service.rs

```diff
@@ -181,74 +181,53 @@ fn spy(
+
         tvu_peers = spy_ref
             .read()
             .unwrap()
             .all_tvu_peers()
             .into_iter()
             .collect::<Vec<_>>();
-        archivers = spy_ref.read().unwrap().all_storage_peers();
+        storage_peers = spy_ref.read().unwrap().all_storage_peers();
+
+        let mut nodes: Vec<_> = tvu_peers.iter().chain(storage_peers.iter()).collect();
+        nodes.sort();
+        nodes.dedup();
+
+        let found_node_by_pubkey = if let Some(pubkey) = find_node_by_pubkey {
+            nodes.iter().any(|x| x.id == pubkey)
+        } else {
+            false
+        };
+
+        let found_node_by_gossip_addr = if let Some(gossip_addr) = find_node_by_gossip_addr {
+            nodes.iter().any(|x| x.gossip == *gossip_addr)
+        } else {
+            false
+        };
+
         if let Some(num) = num_nodes {
-            if tvu_peers.len() + archivers.len() >= num {
-                if let Some(gossip_addr) = find_node_by_gossip_addr {
-                    if tvu_peers
-                        .iter()
-                        .chain(archivers.iter())
-                        .any(|x| x.gossip == *gossip_addr)
-                    {
-                        met_criteria = true;
-                        break;
-                    }
-                }
-                if let Some(pubkey) = find_node_by_pubkey {
-                    if tvu_peers
-                        .iter()
-                        .chain(archivers.iter())
-                        .any(|x| x.id == pubkey)
-                    {
-                        met_criteria = true;
-                        break;
-                    }
-                }
-                if find_node_by_pubkey.is_none() && find_node_by_gossip_addr.is_none() {
+            if nodes.len() >= num {
+                if found_node_by_pubkey || found_node_by_gossip_addr {
                     met_criteria = true;
-                    break;
                 }
-            }
-        }
-        if num_nodes.is_none() {
-            if let Some(pubkey) = find_node_by_pubkey {
-                if tvu_peers
-                    .iter()
-                    .chain(archivers.iter())
-                    .any(|x| x.id == pubkey)
-                {
-                    met_criteria = true;
-                    break;
-                }
-            }
-            if let Some(gossip_addr) = find_node_by_gossip_addr {
-                if tvu_peers
-                    .iter()
-                    .chain(archivers.iter())
-                    .any(|x| x.gossip == *gossip_addr)
-                {
+
+                if find_node_by_pubkey.is_none() && find_node_by_gossip_addr.is_none() {
                     met_criteria = true;
-                    break;
                 }
             }
+        } else if found_node_by_pubkey || found_node_by_gossip_addr {
+            met_criteria = true;
         }
-        if i % 20 == 0 {
+        if met_criteria || i % 20 == 0 {
```

Comment:
I snuck this change in too -- it's confusing sometimes to not see the final view of gossip in the log before the spy node exits